js/dataframe/cell.tsx

```diff
@@ -0,0 +1,279 @@
+import { flexRender } from "@tanstack/react-table";
+import { Cell } from "@tanstack/table-core";
+import React, {
+  FC,
+  ChangeEvent as ReactChangeEvent,
+  ReactElement,
+  FocusEvent as ReactFocusEvent,
+  KeyboardEvent as ReactKeyboardEvent,
+  MouseEvent as ReactMouseEvent,
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+import { useImmer } from "use-immer";
+import { updateCellsData } from "./data-update";
+
+// States
+// # √ Ready
+// # √ Editing
+// # √ Saving / Disabled
+// # √ Error
+// # √ Saved
+// # Cancelled (is Ready state?)
+// # New
+// # Added
+// # Removed
+export const CellStateMap = {
+  EditSaving: "EditSaving",
+  EditSuccess: "EditSuccess",
+  EditFailure: "EditFailure",
+  Editing: "Editing",
+  Ready: "Ready",
+} as const;
+const CellStateClassMap = {
+  EditSaving: "edit-waiting",
+  EditSuccess: "edit-success",
+  EditFailure: "edit-failure",
+  Editing: "editing",
+  Ready: "ready",
+} as const;
+export type CellState = keyof typeof CellStateMap;
+
+interface TableBodyCellProps {
+  id: string | null;
+  cell: Cell<unknown[], unknown>;
+  columns: readonly string[];
+  canEdit: boolean;
+  editRowIndex: number | null;
+  editColumnIndex: number | null;
+  setEditRowIndex: (index: number | null) => void;
+  setEditColumnIndex: (index: number | null) => void;
+  cellEditMap: Map<string, { value: string; state: CellState }>;
+  setData: (fn: (draft: unknown[][]) => void) => void;
+  setCellEditMap: (
+    fn: (draft: Map<string, { value: string; state: CellState }>) => void
+  ) => void;
+  maxRowSize: number;
+}
+
+export const TableBodyCell: FC<TableBodyCellProps> = ({
+  id,
+  cell,
+  columns,
+  canEdit,
+  editRowIndex,
+  editColumnIndex,
+  setEditRowIndex,
+  setEditColumnIndex,
+  cellEditMap,
+  setData,
+  setCellEditMap,
+  maxRowSize,
+}) => {
+  const rowIndex = cell.row.index;
+  const columnIndex = cell.column.columnDef.meta!.colIndex;
+
+  const initialValue = cell.getValue();
+  // We need to keep and update the state of the cell normally
+  const [value, setValue] = useImmer(initialValue);
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  const [cellState, setCellState] = useState<CellState>(CellStateMap.Ready);
+
+  useEffect(() => {
+    const cellIsEditable =
+      editRowIndex === rowIndex && editColumnIndex === columnIndex;
+    if (cellIsEditable) {
+      setCellState(CellStateMap.Editing);
+    }
+  }, [editRowIndex, editColumnIndex, rowIndex, columnIndex]);
+
+  const hasUpdated = cellEditMap.has(`[${rowIndex}, ${columnIndex}]`);
+
+  const tableCellClass = CellStateClassMap[cellState];
+
+  const [errorTitle, setErrorTitle] = useState<string | undefined>(undefined);
+
+  const resetEditInfo = React.useCallback(() => {
```

Comment:
```suggestion
  const resetEditInfo = () => {
```

Reply:
Because `resetEditInfo` is used within a `useEffect` ~ line 217. The exhaustive dependencies linter wants it to be a `useCallback`. 

Leaving as a useCallback to not fight the linter